numpy/random/_generator.pyx

```diff
@@ -4414,7 +4415,11 @@ cdef class Generator:
                 with self.lock, nogil:
                     _shuffle_raw_wrap(&self._bitgen, n, 1, itemsize, stride,
                                       x_ptr, buf_ptr)
-        elif isinstance(x, np.ndarray) and x.ndim and x.size:
+        elif isinstance(x, np.ndarray):
+            if x.ndim == 0 or x.size == 0:
+                # ndim or size is 0, so shuffling is a no-op
```

Comment:
I'd argue shuffling a 0d array is an error, and we should let the error propagate

Reply:
Doing this check after `swapaxes` would be sufficient to rule out bad arguments like `shuffle(array([]), axis=12)`

Reply:
> I'd argue shuffling a 0d array is an error, and we should let the error propagate

That just means users need to special-case empty arrays. The end result is correct (no values got shuffled), so I see no reason to raise an exception here.

Reply:
Empty arrays and 0d arrays are very different things

Reply:
At any rate, this needs changing to not ignore illegal axis arguments just because the array is empty.

Reply:
Turns out 0-D arrays were already raising `TypeError` because of `len(x)` at the top of the function. I added a test for that, and raised `AxisError` for invalid axis. The check for `ndim > 0` that was previously there was doing nothing.